src/lm/diff.cpp

```diff
@@ -0,0 +1,255 @@
+/**
+ * @file diff.cpp
+ * @author Sean Massung
+ */
+
+#include <algorithm>
+#include <queue>
+#include "lm/diff.h"
+#include "porter2_stemmer.h"
+
+namespace meta
+{
+namespace lm
+{
+diff::diff(const cpptoml::table& config) : lm_{config}
+{
+    auto table = config.get_table("diff");
+    if (!table)
+        throw diff_exception{"missing [diff] table from config"};
+
+    auto nval = table->get_as<int64_t>("n-value");
+    if (!nval)
+        throw diff_exception{"n-value not specified in config"};
+    n_val_ = *nval;
+
+    auto edits = table->get_as<int64_t>("max-edits");
+    if (!edits)
+        throw diff_exception{"max-edits not specified in config"};
+    max_edits_ = *edits;
+
+    auto b_pen = table->get_as<double>("base-penalty");
+    base_penalty_ = b_pen ? *b_pen : 0.0;
+
+    auto i_pen = table->get_as<double>("insert-penalty");
+    insert_penalty_ = i_pen ? *i_pen : 0.0;
+
+    auto s_pen = table->get_as<double>("substitute-penalty");
+    substitute_penalty_ = s_pen ? *s_pen : 0.0;
+
+    auto r_pen = table->get_as<double>("remove-penalty");
+    remove_penalty_ = r_pen ? *r_pen : 0.0;
+
+    set_stems(*table);
+    set_function_words(*table);
+}
+
+std::vector<std::pair<sentence, double>>
+    diff::candidates(const sentence& sent, bool use_lm /* = false */)
+{
+    use_lm_ = use_lm;
+    using pair_t = std::pair<sentence, double>;
+    auto comp = [](const pair_t& a, const pair_t& b)
+    {
+        return a.second < b.second;
+    };
+    std::priority_queue<pair_t, std::vector<pair_t>, decltype(comp)> candidates{
+        comp};
+    add(candidates, sent);
+
+    seen_.clear();
+    step(sent, candidates, 0);
+
+    std::vector<pair_t> sorted;
+    while (!candidates.empty())
+    {
+        sorted.emplace_back(std::move(candidates.top()));
+        candidates.pop();
+    }
+    std::reverse(sorted.begin(), sorted.end());
+    return sorted;
+}
+
+template <class PQ>
+void diff::add(PQ& candidates, const sentence& sent)
+{
+    seen_.insert(sent.to_string());
+    auto score = lambda_ * lm_.perplexity_per_word(sent)
+                 + (1.0 - lambda_) * sent.average_weight();
+    candidates.emplace(sent, score);
+    if (candidates.size() > max_cand_size_)
+        candidates.pop();
+}
+
+uint64_t diff::least_likely_ngram(const sentence& sent) const
+{
+    double min_prob = 1;
+    uint64_t best_idx = 0;
+    for (uint64_t i = n_val_; i < sent.size(); ++i)
+    {
+        auto ngram = sent(i - n_val_, i);
+        auto prob = lm_.log_prob(ngram);
+        if (prob < min_prob)
+        {
+            min_prob = prob;
+            best_idx = i - 1;
+        }
+    }
+
+    return best_idx;
```

Comment:
Wait, so what is this returning? The index of the word that ends the least likely ngram?